aiutils/tftools/train.py

```diff
@@ -0,0 +1,167 @@
+ - add_variables(variables, optimizer)
+ - add_variables(variables, learning_rate, [other_params])
+
+Usage 1:
+    a = tf.Variable(1)
+    b = tf.Variable(2)
+    c = tf.Variable(3)
+
+    loss = f(a,b,c)
+
+    optimizer = MultiRateOptimizer(tf.train.GradientDescentOptimizer)
+    optimizer.add_variables([a,b], learning_rate=.1)
+    optimizer.add_variables([c], learning_rate=.01)
+
+    min_opt = optimizer.minimize(loss)
+
+    min_opt.eval()
+
+Usage 2:
+    a = tf.Variable(1)
+    b = tf.Variable(2)
+    c = tf.Variable(3)
+
+    loss = f(a,b,c)
+
+    optimizer = MultiRateOptimizer()
+    optimizer.add_variables([a,b], tf.train.GradientDescentOptimizer(.1))
+    optimizer.add_variables([c], tf.train.GradientDescentOptimizer(.01))
+
+    min_opt = optimizer.minimize(loss)
+
+    min_opt.eval()
+"""
+
+import tensorflow as tf
+import itertools
+
+
+class MultiRateOptimizer():
+    """ Class for managing a multi-rate optimization problem """
+
+    def __init__(self, default_optimizer=None):
+        self.optimizers = []
+        self.variables = []
+        self.default_optimizer = default_optimizer
+
+    def check_variables(self, variables):
+        """ Checks variables against the already added variables and returns a list
+        of reused variables.
+
+        variables (list of tf.variables): The variables to check
+
+        return (list of tf.variables): The variables which are already known to this optimizer.
+        """
+        in_variables = set(itertools.chain(*self.variables))
+
+        dupes = []
+        for v in variables:
+            if v in in_variables:
+                dupes.append(v)
+
+        return dupes
+
+    def add_variables(self,
+                      variables,
+                      optimizer=None,
+                      learning_rate=None,
+                      other_params={}):
+        """ Adds Variables and optimizers with different parameters.
+
+        variables (list of tf.variables): the variables to optimize wrt.
+
+        Either:
+        optimizer (tf.train.Optimizer): The corresponding optimizer.
+        Or:
+        learning_rate (float): A learning rate to pass to the default_optimizer
+        other_params (dict): A dictionary of param_name, value to pass the the default optimizer
+        """
+
+        print 'test'
+        chck_vars = self.check_variables(variables)
+        if len(chck_vars) != 0:
+            raise ValueError('Expected all new variables, got overlap', *
+                             [v.name for v in chck_vars])
+        assert (len(self.check_variables(variables)) == 0)
+        self.variables.append(variables)
+
+        if (optimizer is not None):
+            self.optimizers.append(optimizer)
+        else:
+            if self.default_optimizer is None:
+                raise ValueError(
+                    'default_optimizer is None',
+                    'When optimizer is not passed to add_variables, expect default_optimizer to be not None')
```

Comment:
Rephrase error msg. Maybe -
`default_optimizer can not be None if optimizer is not passed to add_variables`